controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+                    "dequeued a new request while one is already outstanding!",
+                );
+
+                let request = outstanding_request.as_mut().unwrap();
+                match send_outgoing_request(&socket, &peer_addr, request, &mut send_buf).await {
+                    Ok(n) => debug!(log, "sent message"; "message" => ?request, "n_bytes" => n),
+                    Err(e) => error!(log, "failed to send message"; "reason" => ?e),
+                }
+            }
+
+            // If we _do_ have an outstanding request, we need to resend it
+            // periodically until we get a response. Wait for up to the resend
+            // interval of inactivity, and then possibly retry.
+            _ = retry_timeout, if outstanding_request.is_some() => {
+                debug!(log, "timed out without response, retrying");
+                let request = outstanding_request.as_mut().unwrap();
+                send_outgoing_request(&socket, &peer_addr, request, &mut send_buf).await.unwrap();
+                info!(log, "resent message");
+            }
+
+            // Poll for outgoing responses we need to send.
+            maybe_response = outgoing_response_rx.recv() => {
+                let response = match maybe_response {
+                    Some(r) => r,
+                    None => {
+                        debug!(log, "outgoing response channel closed, exiting");
+                        return;
+                    }
+                };
+                // TODO-implement
+                debug!(log, "outgoing response: {response:?}");
+            }
+
+            // Poll for incoming packets.
+            res = socket.recv_from(&mut recv_buf) => {
+                let (n_bytes, peer) = match res {
+                    Err(e) => {
+                        error!(log, "I/O error receiving UDP packet: {e:?}");
+                        continue;
+                    }
+                    Ok((n_bytes, peer)) => (n_bytes, peer),
+                };
+
+                // Deserialize the message itself.
+                let (message, remainder): (Message, _) = match hubpack::deserialize(&recv_buf) {
+                    Err(e) => {
+                        error!(
+                            log,
+                            "failed to deserialize message";
+                            "reason" => ?e,
+                            "peer" => peer,
+                            "n_bytes" => n_bytes,
+                        );
+                        continue;
+                    }
+                    Ok((msg, remainder)) => (msg, remainder),
+                };
+                debug!(log, "message from peer"; "peer" => peer, "message" => ?message);
+
+                // Sanity check the protocol version.
+                if message.header.version != message::version::V1 {
+                    if let Err(e) = send_version_mismatch(&log, &socket, &peer, &message.header, &message.modules).await {
```

Comment:
I think we only want to `send_version_mismatch` if we received a mismatched `HostRequest`; if we received a mismatched `HostResponse`, then something _else_ is wrong, but replying to the SP won't help.

Reply:
Good point, thank you. I'll move this until we've guaranteed we're looking at a request from the SP.

Also, FYI -- These argument signatures are getting a bit long, especially since I'm going to add some DTrace probes. I'm planning to transform the `io_loop` function into a POD type and make `send_version_mismatch` etc inherent methods. Apologies for the churn on this!

Reply:
This is obviated by the other comment thread, pointing out that if the version is wrong we don't want to trust anything about the message. We're opting to just drop it now, and never send version mismatches over the wire.

Reply:
👍🏻